ambry-router/src/main/java/com/github/ambry/router/QuotaAwareOperationController.java

```diff
@@ -0,0 +1,193 @@
+   * @param nonBlockingRouter {@link NonBlockingRouter} object.
+   * @throws IOException if the network components could not be created.
+   */
+  public QuotaAwareOperationController(String suffix, String defaultPartitionClass, AccountService accountService,
+      NetworkClientFactory networkClientFactory, ClusterMap clusterMap, RouterConfig routerConfig,
+      ResponseHandler responseHandler, NotificationSystem notificationSystem, NonBlockingRouterMetrics routerMetrics,
+      KeyManagementService kms, CryptoService cryptoService, CryptoJobHandler cryptoJobHandler, Time time,
+      NonBlockingRouter nonBlockingRouter) throws IOException {
+    super(suffix, defaultPartitionClass, accountService, networkClientFactory, clusterMap, routerConfig,
+        responseHandler, notificationSystem, routerMetrics, kms, cryptoService, cryptoJobHandler, time,
+        nonBlockingRouter);
+    this.nonBlockingRouter = nonBlockingRouter;
+    this.routerMetrics = routerMetrics;
+  }
+
+  @Override
+  protected void pollForRequests(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop) {
+    try {
+      List<RequestInfo> newRequestsToSend = new ArrayList<>();
+      pollNewRequests(newRequestsToSend, requestsToDrop);
+      addToRequestQueue(newRequestsToSend);
+      drainRequestQueue(requestsToSend);
+    } catch (Exception e) {
+      logger.error("Operation Manager poll received an unexpected error: ", e);
+      routerMetrics.operationManagerPollErrorCount.inc();
+    }
+  }
+
+  /**
+   * Poll for new requests from each of the operation managers, and initiate backgroud deletes.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the new requests to be sent out.
+   * @param requestsToDrop a list of correlation IDs that will contain the IDs for requests that the network layer
+   *                       should drop.
+   */
+  private void pollNewRequests(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop) {
+    putManager.poll(requestsToSend, requestsToDrop);
+    getManager.poll(requestsToSend, requestsToDrop);
+    nonBlockingRouter.initiateBackgroundDeletes(getBackGroundDeleteRequests());
+    clearBackGroundDeleteRequests();
+    deleteManager.poll(requestsToSend, requestsToDrop);
+    ttlUpdateManager.poll(requestsToSend, requestsToDrop);
+    undeleteManager.poll(requestsToSend, requestsToDrop);
+  }
+
+  /**
+   * Add the specified {@link List} of {@link RequestInfo}s to the reqeust queue.
+   * @param requestInfos {@link List} of {@link RequestInfo} objects.
+   */
+  private void addToRequestQueue(List<RequestInfo> requestInfos) {
+    for (RequestInfo requestInfo : requestInfos) {
+      QuotaResource quotaResource = requestInfo.getChargeable().getQuotaResource();
+      if (quotaResource == null) {
+        quotaResource = UNKNOWN_QUOTA_RESOURCE;
+      }
+      requestQueue.putIfAbsent(quotaResource, new LinkedList<>());
+      requestQueue.get(quotaResource).add(requestInfo);
```

Comment:
Like we discussed offline, just a note for later changes. Seems like we also have logic in router to drop requests if there is no response for them from server for some time (probably due to closed connections, etc?) . Looks like we may need to account for the time spent by them waiting in request queue.

Reply:
Makes sense. Have added a ticket for that as a follow up (shared with you separately).